### PR TITLE
Schedule: track the waiting tasks with task ID, and deleted the scheduled task with exceeded state from the waiting tasks queue

### PR DESCRIPTION
**Pull request by fzhedu**
… queue

Signed-off-by: fzhedu <fzhedu@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/4954

Problem Summary:
It is a rule that the lifecycle of a task in the minTSO schedule should be covered by the taskManager. 
However, the scheduled tasks due to exceeding the hard limit are not removed from the waiting task queue of the schedule, but will ungester from the taskManager, so the waiting task queue hold the last reference of the task, which break the rule.

to fundamentally prevent from the probem above, we track the waiting tasks with its task ID.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Mpp/MinTSOScheduler.cpp [hunk 1]
```diff
@@ -153,9 +155,12 @@ void MinTSOScheduler::scheduleWaitingQueries(MPPTaskManager & task_manager)
         /// schedule tasks one by one
         while (!query_task_set->waiting_tasks.empty())
         {
-            auto task = query_task_set->waiting_tasks.front();
-            if (!scheduleImp(current_query_id, query_task_set, task, true))
+            auto task = query_task_set->task_map.find(query_task_set->waiting_tasks.front())->second;
+            if (task != nullptr && !scheduleImp(current_query_id, query_task_set, task, true))
+            {
+                query_task_set->waiting_tasks.pop(); /// it should be pop from the waiting queue, as the task is scheduled as the exceeded state.
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/MinTSOScheduler.cpp:161]: Seems we can not remove the waiting task for all the cases, consider that a task is not min-tso, and it is not scheduled due to soft limit, then why remove it from `waiting_tasks`?

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Flash/Mpp/MinTSOScheduler.cpp [hunk 2]
```diff
@@ -97,7 +97,9 @@ void MinTSOScheduler::deleteQuery(const UInt64 tso, MPPTaskManager & task_manage
         {
             while (!query_task_set->waiting_tasks.empty())
             {
-                query_task_set->waiting_tasks.front()->scheduleThisTask(MPPTask::ScheduleState::FAILED);
+                auto task = query_task_set->task_map.find(query_task_set->waiting_tasks.front())->second;
```
[the remark this answers is not in the document]
Comment (bestwoody) [reply dbms/src/Flash/Mpp/MinTSOScheduler.cpp:100]: +1

File: dbms/src/Flash/Mpp/MinTSOScheduler.cpp [hunk 3]
```diff
@@ -153,9 +155,12 @@ void MinTSOScheduler::scheduleWaitingQueries(MPPTaskManager & task_manager)
         /// schedule tasks one by one
         while (!query_task_set->waiting_tasks.empty())
         {
-            auto task = query_task_set->waiting_tasks.front();
-            if (!scheduleImp(current_query_id, query_task_set, task, true))
+            auto task = query_task_set->task_map.find(query_task_set->waiting_tasks.front())->second;
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply dbms/src/Flash/Mpp/MinTSOScheduler.cpp]: fixed

**Review by bestwoody — APPROVED** [review]
lgtm

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**